examples/tp/tm_high_order.py

```diff
@@ -0,0 +1,465 @@
+print "predicts B, B predicts C, but this time C outputs a simultaneous prediction of"
+print "both D and Y. In order to disambiguate, the TM would require to know if the"
+print "preceding characters were AB or XB. When presented with character X the TM"
+print "predicts B, whereas Y and D yield no prediction."
+print "-"*50
+print ""
+
+showPredictions()
+
+# PART 3. Now we will present noisy inputs to the TM. We will add noise to the sequence XBCY
+# by corrupting 30% of its bits. We would like to see how the TM responds in the presence of
+# noise and how it recovers from it.
+print ""
+print "-"*50
+print "Part 3. We will add noise to the sequence XBCY by corrupting 30% of the bits in the vectors"
+print "encoding each character. We would expect to see a decrease in prediction accuracy as the"
+print "TM is unable to learn the random noise in the input (Fig 3). However, this decrease is not"
+print "significant."
+print "-"*50
+print ""
+
+x = []
+y = []
+trainTM(seq2, timeSteps=50, noiseLevel=0.3)
+
+plt.ylim([-0.1,1.1])
+plt.plot(x, y)
+plt.xlabel("Timestep")
+plt.ylabel("Prediction Accuracy")
+plt.title("Fig. 3: Accuracy in TM with 30% noise in input")
+plt.savefig("figure_3")
+plt.close()
+
+print ""
+print "-"*50
+print "Let's have a look again at the output of the TM when presented with noisy"
+print "input (30%). Here, the noise is low so the TM does not try to learn the noise,"
+print "which would be the case if we saw 'noisy' columns being predicted when"
+print "presented with individual characters. Thus, we could say that the TM exhibits"
+print "resilience to noise in its input."
+print "-"*50
+print ""
+
+showPredictions()
+
+# Let's corrupt the sequence more by adding 50% of noise to each of its characters.
+# Here, we would expect to see some 'noisy' columns being predicted when the TM is
+# presented with the individual characters.
+print ""
+print "-"*50
+print "Now, we will 50% of the bits in the characters X, B, C, and Y. This time we will"
+print "corrupt 90% of its contents. As expected, the accuracy will decrease (Fig 5) and"
+print "'noisy' columns will be predicted by the TM."
+print "-"*50
+print ""
+
+x = []
+y = []
+trainTM(seq2, timeSteps=50, noiseLevel=0.5)
+
+print ""
+print "-"*50
+print "Let's have a look again at the output of the TM when presented with noisy"
+print "input. The prediction of some characters (eg. X) now includes columns that"
+print "are not related to any other character. This is because the TM tried to learn"
+print "the noise in the input patterns."
+print "-"*50
+print ""
+
+showPredictions()
+
+plt.ylim([-0.1,1.1])
+plt.plot(x, y)
+plt.xlabel("Timestep")
+plt.ylabel("Prediction Accuracy")
+plt.title("Fig. 4: Accuracy in TM with 50% noise in input")
+plt.savefig("figure_4")
+plt.close()
+
+# Will the TM be able to forget the 'noisy' columns learned in the previous step?
+# We will present the TM with the original sequence XBCY in order to re-learn it.
```

Comment:
The terminology is a bit confusing. The TM is not really re-learning the sequence - it never forgot it. What you are doing here is forgetting the noise which it also learned.